### PR TITLE
Update jacobian to include energy from water fluxes

### PR DESCRIPTION
**Pull request by juliasloan25**
## Purpose 
To improve the stability of our longruns, we want to include the `dT_rhoe/dvartheta_l` term in our jacobian approximation. We have found that, in the case of high water flux near saturation, the `dpsi/dtheta` term that this depends on becomes very large and unstable to step explicitly. To demonstrate that the water flux energy is the culprit, we tried removing this term from the explicit tendency and saw improved stability.

See #953 for plots justifying the need for this change

## To-do
- [x] update jacobian construction to accommodate off-diagonal terms
- [x] update jacobian calculation to include `dT_rhoe/dvartheta_l` term for `EnergyHydrology` model

### CODE REVIEW

File: experiments/standalone/Soil/energy_hydrology_saturated.jl [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+set_initial_cache!(p, Y, t0);
+sol_dt_small = SciMLBase.solve(prob, ode_algo; dt = 100.0);
+init_soil!(Y, z, soil.parameters);
+set_initial_cache!(p, Y, t0);
+sol_dt_large = SciMLBase.solve(prob, ode_algo; dt = 900.0);
+@test sum(isnan.(sol_dt_large.u[end])) == 0
+
+norm(x) = sqrt(mean(parent(x .^ 2)))
+rmse(x, y) = sqrt(mean(parent((x .- y) .^ 2)))
+@test rmse(sol_dt_small[end].soil.ϑ_l, sol_dt_large[end].soil.ϑ_l) /
```
Comment (kmdeck) [inline experiments/standalone/Soil/energy_hydrology_saturated.jl:133]: @test vs @assert?

**Review by kmdeck — APPROVED** [review]

File: experiments/standalone/Soil/energy_hydrology_saturated.jl (hunk 1 above)
Comment (juliasloan25) [reply experiments/standalone/Soil/energy_hydrology_saturated.jl:133]: maybe it's better to use assert, and then we can remove the Test dependency in this file

File: src/shared_utilities/implicit_timestepping.jl [hunk 2]
```diff
@@ -156,12 +171,22 @@ function ImplicitEquationJacobian(Y::ClimaCore.Fields.FieldVector)
...
+        explicit_blocks...,
+    )
 
-
-    matrix = MatrixFields.FieldMatrix(implicit_blocks..., explicit_blocks...)
-
-    # Set up block diagonal solver for block Jacobian
-    alg = MatrixFields.BlockDiagonalSolve()
+    # Choose algorithm based on whether off-diagonal blocks are present
+    # TODO change this to use dispatch on model type
```
Comment (juliasloan25) [inline src/shared_utilities/implicit_timestepping.jl:181]: Can we delete this TODO for now and just keep track of it in an issue? (I'll open one)

File: experiments/standalone/Soil/energy_hydrology_saturated.jl [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+)
+Y, p, cds = initialize(soil);
+
+set_initial_cache! = make_set_initial_cache(soil)
+exp_tendency! = make_exp_tendency(soil)
+imp_tendency! = make_imp_tendency(soil);
+jacobian! = ClimaLand.make_jacobian(soil);
+jac_kwargs = (; jac_prototype = ImplicitEquationJacobian(Y), Wfact = jacobian!);
+
+
```
Comment (juliasloan25) [inline experiments/standalone/Soil/energy_hydrology_saturated.jl:73]:
```suggestion
```

File: src/standalone/Soil/energy_hydrology.jl [hunk 4]
```diff
@@ -373,6 +374,28 @@ function ClimaLand.make_compute_jacobian(model::EnergyHydrology{FT}) where {FT}
...
+                        hydrology_cm,
+                        Y.soil.ϑ_l,
+                        ν - Y.soil.θ_i, #ν_eff
+                        θ_r,
+                        S_s,
+                    ),
+                )
+            ) - (I,)
+
+
```
Comment (juliasloan25) [inline src/standalone/Soil/energy_hydrology.jl:397]:
```suggestion
```

File: src/shared_utilities/implicit_timestepping.jl (hunk 2 above)
Comment (kmdeck) [reply src/shared_utilities/implicit_timestepping.jl:181]: thank you!

**Pull request merged**